data/blog/2024-10-18-react-server-components.mdx

```diff
@@ -0,0 +1,230 @@
+
+### When should you use  **React Server Components**?
+
+**You should use React Server Components when**:
+
+- You have static, non-interactive content that doesn’t change frequently.
+- You want to improve performance by reducing JavaScript sent to the client.
+- SEO is important for your application.
+- You want to offload heavy data-fetching or computational tasks to the server.
+
+**You shouldn’t React Server Components when**:
+
+- Your components require client-side interactivity, state, or hooks.
+- You need access to browser APIs or require real-time updates.
+- You are building complex pages with heavy client-side logic or user-specific content.
+
+### Common Pitfalls
+
+**Client Components are not rendered entirely on the Client**
+
+They are rendered both on Client and Server. They are named Client Components to differentiate them from Server Components.
+
+**`‘use server’` directive is for Server Actions**
+
+By default, Next.js uses Server Components so you don’t need to specify anything. If you want to use Client Components, you add the `‘use client’` directive to declare a boundary between Server and Client modules. `‘use server’` is an entirely different directive used for **Server Actions**, which is beyond the scope of this blogpost.
+
+## Project Setup
+
+We will use the Blog template from Vercel:
+
+```bash
+pnpm create next-app --example https://github.com/vercel/examples/tree/main/solutions/blog blog
+```
+
+You can run the example locally and see what it looks like:
+
+```bash
+cd blog
+pnpm dev
+```
+
+Let’s install `@upstash/redis`:
+
+```bash
+pnpm add @upstash/redis
+```
+
+## Environment Setup
+
+1. Go to [Upstash Console → Redis](https://console.upstash.com/redis) and create a new Database:
+
+![Create Database](/blog/react-server-components/database-create.png)
+
+2. Scroll down to REST API section, switch to `.env` tab and copy the environment variables for the next step:
+
+![Upstash Environment Variables](/blog/react-server-components/database-env.png)
+
+3. Create a `.env` file and paste your environment variables:
+    
+```bash
+UPSTASH_REDIS_REST_URL=<YOUR_URL>
+UPSTASH_REDIS_REST_TOKEN=<YOUR_TOKEN>
+```
+    
+## Setup Views Component
+
+Create `/app/components/views.tsx`:
+
+```tsx
+import { headers } from 'next/headers'
+import { Redis } from "@upstash/redis"
+
+const redis = Redis.fromEnv();
+
+async function view(slug: string, ip: string) {
+    // Hash the IP address to anonymize it
+    const buf = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(ip));
+    const hash = Array.from(new Uint8Array(buf)).map((b) => b.toString(16).padStart(2, "0")).join("");
+    // Deduplicate views
+    const newView = await redis.set(`deduplicate:${hash}:${slug}`, true, {
+        nx: true, // Only set the key if it doesn't exist
+        ex: 24 * 60 * 60, // Expire the key after 24 hours
+    });
+    if (newView) {
+        await redis.incr(`pageviews:${slug}`); // Increment the view count
```

Comment:
incr actually returns the new value of the key after incrementing. So we can handle it all in a single redis command

Reply:
`incr` is ran conditionally here. I think current design is more simple and readable but I can change if you think this is better: 
```
...
if (newView) {
        return await redis.incr(`pageviews:${slug}`);
} else {
        return await redis.get<number>(`pageviews:${slug}`) || 0
}
...
const views = await view(slug, ip)
...
 ```